### ISSUE TITLE
Alternatives to NativeMessaging

### ISSUE THREAD

**Issue by jelmervdl** [XapaJIaMnu/translateLocally#117]
The way NativeMessaging works doesn't seem to work well with all these sandboxing and packaging efforts that are happening. In at least the macOS sandbox, and with flatpack/snap packages the application (i.e. translateLocally) doesn't have complete access to the filesystem. So it can't register itself with browsers. Similiarly, browsers could be shipped through those methods, and then don't have access to running translateLocally.

Issues for reference:
1. https://github.com/flatpak/xdg-desktop-portal/issues/655
2. https://developer.apple.com/forums/thread/92554

So… I'm looking for alternatives:
1. https://textslashplain.com/2019/08/28/browser-architecture-web-to-app-communication-overview/
3. https://wicg.github.io/private-network-access/
4. Running a WebRTC locally that's only visible from localhost?
5. … all of the above need some daemon running in the background that's started separately

**Comment by XapaJIaMnu** [comment]
@kpu as far as I recall the original implementation was indeed webrtc server listening on localhost and we went native messaging to avoid it XD...

**Comment by kpu** [comment]
WebRTC the video / voice call standard?  
We used to have a REST server but it was janky, particularly with regard to security.  https://www.zdnet.com/article/zoom-defends-use-of-local-web-server-on-macs-after-security-report/

**Comment by jelmervdl** [comment]
Yep, but WebRTC can also [just use the data channel](https://simpl.info/rtcdatachannel/). The useful part is the peer-to-peer bit because I think you can use it to set up a connection to something running on localhost without having to issue a certificate for localhost https ([because that's a bad idea](https://letsencrypt.org/docs/certificates-for-localhost:))

Chrome published a bit more info about how they're treating/tightening local access, but what I'm getting from it is that if you implement a server [to correctly respond to preflight requests](https://developer.chrome.com/blog/private-network-access-preflight/), 127.0.0.1 will still work? I can't really find info. Normally fetching something from http from a https orgin is not allowed because it is potentially unsafe. But I think 127.0.0.1 is always seen as secure because there is no middle to mitm.

Ugh… I can't find any documentation on it. But a simple test, running `nc -l 8000`, going to `https://translatelocally.com` and typing `fetch('http://127.0.0.1:8000/', {mode: 'cors'})` in the web inspector console does show Chrome at least making an attempt at doing preflight checks so I'm hopeful the whole webRTC dance isn't necessary and we could just run a simple HTTP server on localhost.

**Comment by jelmervdl** [comment]
I suppose this is all exactly [how bergamot was initially doing things, hahaha](https://github.com/browsermt/mts)
